### PR TITLE
test(unit): add comprehensive unit test suites for core library classes & lib

### PR DESCRIPTION
**Pull request by Shadow243**
####  New Test Suites

- tests/phpunit/modules/core/mailbox.php – Tests for Hm_Mailbox
  - Bridge pattern for IMAP, JMAP, EWS, SMTP
  - Connection/authentication handling
  - Search across protocols
  - Error handling and edge cases

- tests/phpunit/lib/searchable.php – Tests for Searchable trait
  - getBy() method with parameters and filters
  - Edge cases and errors

- tests/phpunit/lib/scram_authenticator.php – SCRAM authentication
  - Hash algorithm detection/validation
  - Client proof for multiple algorithms
  - Authentication flow

- tests/phpunit/lib/ini_set.php – PHP configuration validation
  - Session security, compression, open_basedir
  - Compatibility: PHP 7.3+, 8.0+, 8.1

### CODE REVIEW

File: tests/phpunit/lib/searchable.php [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+use PHPUnit\Framework\TestCase;
+
+/**
+ * Tests for the Searchable trait
+ */
+class Hm_Test_Searchable extends TestCase {
+
+    public function setUp(): void {
+        if (!defined('APP_PATH')) {
+            require_once __DIR__.'/../bootstrap.php';
```
Comment (kroky) [inline tests/phpunit/lib/searchable.php]: Proper bootstrap means we don't call these inside the unit tests themselves. Why would the app path not be defined? Tests should be executed by a single phpunit command which should always properly bootstrap the framework.

File: tests/phpunit/lib/searchable.php [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+        }
+        // Load mocks first
+        if (!class_exists('Hm_Mock_Session', false)) {
+            require_once APP_PATH.'tests/phpunit/mocks.php';
```
Comment (kroky) [inline tests/phpunit/lib/searchable.php]: This is not the correct place to include such files. Do it in the bootstrap system.

File: tests/phpunit/mocks.php [hunk 3]
```diff
@@ -189,6 +189,80 @@ public function __construct($type) {
         $this->type = $type;
     }
 }
+
+trait Mock_Searchable {
```
Comment (kroky) [inline tests/phpunit/mocks.php]: None of the classes or traits called mocks are actual mocks - you copy/implement details from the actual code in the test suite - this is counterproductive as code will diverge over time. See the documentation on how to mock objects https://docs.phpunit.de/en/9.6/test-doubles.html#mock-objects  - you add some expectations to methods that should be called and what should be returned but you actually work with the real objects and not some classes that resemble the real classes partially. In theory, it is best to follow this pattern:
- use the object itself whenever possible
- if testing becomes complicated as the original class depends on 3rd party external connections, you can mock the result of some of its methods and return what you want to test
- note that merely testing the return of data that you supplied as a mock is not a productive test. The test objective is to test code that runs in the framework and not code that is written in the tests themselves. So, whenever you want to test a method, see what external connections it uses, mock them and actually call that method on that original class/object and test that it works correctly. There is a subtle but important difference here. The way you did it with these mocks is that you test the mocks themselves which doesn't make sense.

Also, you shouldn't use reflection to set private properties or do similar changes that were not designed to be done with the class in question. If you want to test something that's not available for some reason, do:
- think if it actually needs to be tested. Private methods usually are not tested. Testing the public interface of the class is enough to ensure it works fine as it calls private methods internally, so they become part of the tests.
- if something private really needs to be tested, refactor the original class/interface and make it public or separate it to parts that are testable
Refactoring original code because of testing is a legitimate use-case, so don't be afraid to do it.

Comment (Shadow243) [reply tests/phpunit/mocks.php]: @kroky I’ve applied all the requested corrections and fixed the other failing tests.

For Hm_Test_Mailbox, I added some imports in setUp to avoid conflicts. These are only required there and not elsewhere.

Comment (Shadow243) [reply tests/phpunit/mocks.php]: In addition, in hm-jmap.php I added a few missing compatibility functions (is_supported, get_message_sort_order, and sort_by_fetch) with TODOs and comments. These provide placeholders so we can later implement proper JMAP behavior and confirm everything works correctly with JMAP as well.

File: modules/imap/hm-ews.php [hunk 4]
```diff
@@ -25,1130 +25,1132 @@
  * public interface to EWS mailboxes
  * @subpackage imap/lib
  */
-class Hm_EWS {
-    protected $ews;
-    protected $api;
-    protected $authed = false;
+if (!class_exists('Hm_EWS')) {
```
Comment (kroky) [inline modules/imap/hm-ews.php]: Please remove this check. Require the file just ones and it will be good. You changed the whole file indentation because of this check when PHP already has a mechanism for including class files only once.

Comment (Shadow243) [reply modules/imap/hm-ews.php]: Done

**Comment by IrAlfred** [comment]
This is a very big pull request and affects too many modules. A more iterative approach (closing this and submitting smaller) module-specific PRs would likely lead to faster feedback and safer merges. What's your thoughts @kroky @Shadow243 ?

**Comment by kroky** [comment]
I prefer we do all the related/intended work on the tests here in one PR even if it becomes big. It is easier to maintain from my point of view.

**Comment by Shadow243** [comment]
> I prefer we do all the related/intended work on the tests here in one PR even if it becomes big. It is easier to maintain from my point of view.

Noted, Work in progress.

**Comment by Shadow243** [comment]
@kroky please take a look at this again.

File: tests/phpunit/lib/api_curl.php [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+ * Unit tests for Hm_API_Curl
+ * @package lib/tests
+ */
+
+use PHPUnit\Framework\TestCase;
+
+class Hm_Test_API_Curl extends TestCase {
+
+    public function setUp(): void {
+        require_once __DIR__.'/../bootstrap.php';
```
Comment (kroky) [inline tests/phpunit/lib/api_curl.php]: This should load for all unit tests during bootstrap process not within each individual unit test.

File: tests/phpunit/lib/repository.php [hunk 6]
```diff
@@ -0,0 +1,260 @@
+<?php
+
+/**
+ * Unit tests for Hm_Repository trait
+ * @package lib/tests
+ */
+
+use PHPUnit\Framework\TestCase;
+
+class Hm_Test_Repository extends TestCase {
```
Comment (kroky) [inline tests/phpunit/lib/repository.php:10]: Why do you call it repository test when you test only the taggs wrapper?

**Comment by kroky** [comment]
Overall, it looks good and can be merged after we resolve the 2 issues.

File: tests/phpunit/lib/api_curl.php (hunk 5 above)
Comment (Shadow243) [reply tests/phpunit/lib/api_curl.php]: Done

File: tests/phpunit/lib/repository.php (hunk 6 above)
Comment (Shadow243) [reply tests/phpunit/lib/repository.php:10]: I applied required changes.
